de1plus/profile.tcl

```diff
@@ -513,5 +519,282 @@ namespace eval ::profile {
+            append profile_path ".tcl"
+        }
+        if { ![file exists $profile_path] } {
+            msg -WARNING [namespace current] load_profile: "profile file '$profile_path' not found"
+            return
+        }
+        
+        msg -INFO [namespace current] read_profile: "reading file '$profile_path'"
+        array set profile {}
+    
+        # Set defaults in case they are not defined in the profile file 
+        set profile(final_desired_shot_volume_advanced_count_start) 0
+        set profile(settings_profile_type) "settings_2a"
+        set profile(beverage_type) "espresso"
+    
+        try {
+            array set arrfile [encoding convertfrom utf-8 [read_binary_file $profile_path]]
+        } on error err {
+            msg -WARNING [namespace current] read_profile: "error reading profile '$fn': $err"
+            return
+        }
+    
+        foreach k [concat [::profile_vars] $extra_vars] {
+            if { [info exists arrfile($k)] } {
+                set profile($k) $arrfile($k)
+            } elseif { ![info exists profile($k)] } {
+                set profile($k) {}
+            }
+        }
+        
+        set profile(settings_profile_type) [fix_profile_type $profile(settings_profile_type)]
+        
+        if { $profile(settings_profile_type) eq "settings_2c" && $profile(final_desired_shot_weight) ne "" &&
+                $profile(final_desired_shot_weight_advanced) eq "" } {
+            msg -NOTICE [namespace current] load_profile: "Using a default for final_desired_shot_weight_advanced " \
+                "from final_desired_shot_weight of $profile(final_desired_shot_weight)"
+            set profile(final_desired_shot_weight_advanced) $profile(final_desired_shot_weight)
+        }
+    
+        # pre-set the shot volume, to the shot weight, if importing an old shot definition that doesn't have a an end volume 
+        if { $profile(final_desired_shot_volume) eq "" } {
+            msg -NOTICE [namespace current] load_profile: "pre-set the shot volume, to the shot weight, if importing an old shot definition" \
+                "that doesn't have a an end volume"
+            set profile(final_desired_shot_volume) $profile(final_desired_shot_weight)
+        }
+    
+        if { $profile(advanced_shot) eq {} } {
+            # Ensure the profile's advanced_shot variable is always defined
+            switch $profile(settings_profile_type) \
+                settings_2a {
+                    array set temp_profile [pressure_to_advanced_list profile]
+                    set profile(advanced_shot) $temp_profile(advanced_shot)
+                } settings_2b {
+                    array set temp_profile [flow_to_advanced_list profile]
+                    set profile(advanced_shot) $temp_profile(advanced_shot)
+                }
+        }
+        
+        return [array get profile]
+    }
+    
+    # Returns a dictionary with a textual representation of the provided profile.
```

Comment:
would it make sense to convert to json-compat advanced list first and then work on that? That way we at least dont add dependencies onto the TCL profiles

Reply:
I thought JSON versions were mainly for exporting, not for internal processing, or that was only for shots? Shouldn't be too hard to change, but I would leave that for later.

Reply:
I would prefer to remove the TCL profiles in the future, but as long as we are forced to deal with GUI elements tied to ::settings entries it will most likely not be a realistic goal to get the migration up and running smoothly.

Reply:
Ok, sorry I didn't realize to do it directly with the new vesion. If needed I'll be happy to migrate it.